eureka/niriss_class.py

```diff
@@ -0,0 +1,516 @@
+           is None. If None, will search the current working directory
+           for the files.
+        output_dir : str, optional
+           The path where output files will be saved. Default is None.
+           if None, will save all files to the current working directory.
+
+        Attributes
+        ----------
+        filename : str
+           The science file name.
+        data_path : str
+           The path to where the science file is stored.
+        output_dir : str
+           The path to where output files will be stored.
+        """
+        self.filename = filename
+
+        if data_dir is not None:
+            self.data_dir = data_dir
+        else:
+            self.data_dir = os.getcwd()
+
+        if output_dir is not None:
+            self.output_dir = output_dir
+        else:
+            self.output_dir = os.getcwd()
+
+        # Opens the science FITS file and sets up all proper
+        #   data attributes for the reduction steps.
+        self.setup()
+
+        self.dq_mask = data_quality_mask(self.dq)
+
+        if f277_filename is not None:
+            self.setup_f277(f277_filename)
+        else:
+            self.f277 = None
+            print('Without F277W filter image, some functions may not be available.')
+
+        self.trace_ear     = None
+        self.trace_edge     = None
+        self.bkg      = None
+        self.box_mask = None
+        self.box_var1 = None
+        self.box_var2 = None
+        self.bkg_removed  = None
+        self.box_spectra1 = None
+        self.box_spectra2 = None
+        self.box_mask_separate = None
+
+        return
+
+
+    def setup(self):
+        """
+        Sets up all proper attributes from the FITS file.
+
+        Attributes
+        ----------
+        mhdr : FITS header
+           The main header in the FITS file.
+        shdr : FITS header
+           The header associated with the science frame
+           extension.
+        intend : float
+        time : np.ndarray
+           Array of time values based on the exposure start
+           and stop times indicated in the FITS header.
+        time_units : str
+           The units `self.time` is in.
+        inttime : float
+           The effective integration time (in seconds).
+        data : np.ndarray
+           3D array of science group images.
+        err : np.ndarray
+           3D array of errors associated with the science
+           frames.
+        dq : np.ndarray
+           3D array of data quality masks, indicating the
+           location of bad pixels.
+        var : np.ndarray
+           3D array of poisson estimated variances.
+        v0 : np.ndarray
+           3D array of rnoise estimated variances.
+        meta : np.ndarray
+           The array of ASDF additional meta data within
+           the FITS file.
+        median : np.ndarray
+           Median frame of all science images.
+        """
+        hdu = fits.open(os.path.join(self.data_dir, self.filename))
```

Comment:
Most of this code is a copy of the read function and should just call that function instead of duplicating code (which makes it very challenging to maintain)